pkg/kubelet/certificate/certificate_manager.go

```diff
@@ -0,0 +1,244 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package certificate
+
+import (
+	"crypto/ecdsa"
+	"crypto/elliptic"
+	cryptorand "crypto/rand"
+	"crypto/rsa"
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/pem"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/golang/glog"
+
+	"k8s.io/apimachinery/pkg/util/wait"
+	certificates "k8s.io/kubernetes/pkg/apis/certificates/v1beta1"
+	clientcertificates "k8s.io/kubernetes/pkg/client/clientset_generated/clientset/typed/certificates/v1beta1"
+	"k8s.io/kubernetes/pkg/kubelet/util/csr"
+)
+
+const (
+	syncPeriod = 1 * time.Hour
+)
+
+// Manager maintains and updates the certificates in use by this kubelet. In
+// the background it communicates with the API server to get new certificates
+// for certificates about to expire.
+type Manager interface {
+	// Start the API server status sync loop.
+	Start()
+	// GetCertificate gets the current certificate from the certificate
+	// manager. This function matches the signature required by
+	// tls.Config.GetCertificate so it can be passed as TLS configuration. A
+	// TLS server will automatically call back here to get the correct
+	// certificate when establishing each new connection.
+	GetCertificate(clientHello *tls.ClientHelloInfo) (*tls.Certificate, error)
+}
+
+// Store is responsible for getting and updating the current certificate.
+// Depending on the concrete implementation, the backing store for this
+// behavior may vary.
+type Store interface {
+	Current() (*tls.Certificate, error)
+	// Accepts the PEM data for the cert/key pair and makes the new cert/key
+	// pair the 'current' pair, that will be returned by future calls to
+	// Current().
+	Update(cert, key []byte) (*tls.Certificate, error)
+}
+
+type manager struct {
+	certSigningRequestClient clientcertificates.CertificateSigningRequestInterface
+	template                 *x509.CertificateRequest
+	usages                   []certificates.KeyUsage
+	certStore                Store
+	certAccessLock           sync.RWMutex
+	cert                     *tls.Certificate
+	shouldRotatePercent      int32
+}
+
+// NewManager returns a new certificate manager. A certificate manager is
+// responsible for being the authoritative source of certificates in the
+// Kubelet and handling updates due to rotation.
+func NewManager(
+	certSigningRequestClient clientcertificates.CertificateSigningRequestInterface,
+	template *x509.CertificateRequest,
+	usages []certificates.KeyUsage,
+	certificateStore Store,
+	certRotationPercent int32) (Manager, error) {
+
+	cert, err := certificateStore.Current()
+	if err != nil {
+		return nil, err
```

Comment:
Should we instead return an error from GetCertificates() until the store has synced?